backend_plugin/src/main/evaluators/TrailingWhiteSpaceEvaluator.java

```diff
@@ -0,0 +1,68 @@
+package main.evaluators;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.DocumentEvent;
+import org.eclipse.jface.text.IDocument;
+
+/**
+ * Evaluates DocumentEvent changes to determine if the user is manually deleting trailing white spaces
+ * Eclipse has a setting to automatically remove trailing white spaces on save, so the user could use
+ * this instead.
+ */
+public class TrailingWhiteSpaceEvaluator extends FeatureEvaluator {
+	
+	private String lineBeforeChange;
+	
+	/**
+	 * Constructs a TrailingWhiteSpaceEvaluator
+	 * @param document
+	 */
+	public TrailingWhiteSpaceEvaluator(IDocument document) {
+		this.featureID = "trailingWhiteSpaceSuggestion";
+		this.document = document;
+	}
+
+	/**
+	 * Evaluates document changes to see if the user is manually deleting all trailing white space
+	 * from a line
+	 * @param event
+	 */
+	@Override
+	public boolean evaluateDocumentChanges(DocumentEvent event) {
+		
+		try {
+			int line = document.getLineOfOffset(event.getOffset());
+			int startOffset = document.getLineOffset(line);
+			int length = document.getLineLength(line);
+			
+			// Need to check up to length - 1 instead of length due to line feed char at end
+			String lineAfterChange = document.get(startOffset, length - 1);
+			
+			// Check that only the line's whitespace has changed, and that the line no longer
+			// ends with whitespace
+			boolean nonWhiteSpaceCharactersUnchanged = lineBeforeChange.trim().equals(lineAfterChange.trim());
```

Comment:
Is there a way we can limit the amount of times we make these three checks. We don't need to do it on every document change.

If you remove something, then the event.getText().length() == 0, yeah? Can we do that quick check first? That will quickly fail on many of the doc changes that we don't care about here.

Reply:
Good point, I added the check for the event text being of length 0.